### ISSUE TITLE
NotImplementedError: Cannot access storage of SparseTensorImpl

### ISSUE THREAD

**Issue by gcramer23** [pytorch/pytorch#58755]
## 🐛 Bug

when using cuda rpc and returning a sparse tensor to update gradients during the .backward() call `NotImplementedError: Cannot access storage of SparseTensorImpl` is thrown.

## To Reproduce

run the program example below with cuda_rpc=True in `if __name__ == "__main__":`

```python
import os

import torch
import torch.distributed as c10d
import torch.distributed.rpc as rpc
import torch.multiprocessing as mp
import torch.nn as nn
import torch.nn.functional as F
from torch.distributed.rpc import TensorPipeRpcBackendOptions
from torch.nn.parallel import DistributedDataParallel as DDP


def sparse_tensor_to_rpc_format(sparse_tensor):
    sparse_tensor = sparse_tensor.coalesce()
    return [sparse_tensor.indices(), sparse_tensor.values(), sparse_tensor.size()]


def sparse_rpc_format_to_tensor(sparse_rpc_format):
    return torch.sparse_coo_tensor(
        sparse_rpc_format[0], sparse_rpc_format[1], sparse_rpc_format[2]
    ).coalesce()


class Server:

    @staticmethod
    @rpc.functions.async_execution
    def identity(tensor):
        fut = torch.futures.Future()
        fut.set_result(tensor)
        return fut


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.embedding = nn.EmbeddingBag(10, 10, sparse=True)

    def forward(self, x):
        return self.embedding(x)


def run_trainer(rank, rref, cuda_rpc):

    def basic_hook(state, bucket):

        def callback(fut):
            tensor = fut.wait()
            if type(tensor) is list:
                tensor = sparse_rpc_format_to_tensor(tensor)
            if not cuda_rpc:
                tensor = tensor.cuda(rank)
            return [tensor]

        tensor = bucket.get_tensor()
        if not cuda_rpc:
            tensor = tensor.cpu()
        if tensor.is_sparse:
            tensor = sparse_tensor_to_rpc_format(tensor)
        fut = rref.rpc_async().identity(tensor).then(callback)
        return fut

    model = Model().cuda(rank)
    store = c10d.FileStore("/tmp/tmpn_k_8so02", 1)
    process_group = c10d.ProcessGroupGloo(store, rank, 1)
    ddp_model = DDP(model, device_ids=[rank], process_group=process_group)
    ddp_model.register_comm_hook(None, basic_hook)
    loss_fn = nn.MSELoss()
    input = torch.randint(5, (10, 10)).cuda(rank)
    loss_fn(ddp_model(input), input.to(torch.float)).backward()


def run_test(rank, cuda_rpc):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '29501'
    opts = TensorPipeRpcBackendOptions()
    if rank == 2:
        rpc.init_rpc(
            "master",
            rank=rank,
            world_size=3,
            rpc_backend_options=opts
        )
        rref = rpc.remote("server", Server)
        rpc.rpc_async(
            "trainer",
            run_trainer,
            args=(0, rref, cuda_rpc,)
        )
    elif rank == 0:
        if cuda_rpc:
            opts.set_device_map("server", {rank: 1})
        rpc.init_rpc(
            "trainer",
            rank=rank,
            world_size=3,
            rpc_backend_options=opts
        )
    else:
        rpc.init_rpc(
            "server",
            rank=rank,
            world_size=3,
            rpc_backend_options=opts
        )
    rpc.shutdown()


if __name__ == "__main__":
    cuda_rpc = True
    mp.spawn(
        run_test,
        nprocs=3,
        args=(cuda_rpc,),
        join=True
    )
```
```
On WorkerInfo(id=0, name=trainer):
NotImplementedError('Cannot access storage of SparseTensorImpl',)
Traceback (most recent call last):
  File "/fsx/users/gcramer/work/pytorch/torch/distributed/rpc/internal.py", line 210, in _run_function
    result = python_udf.func(*python_udf.args, **python_udf.kwargs)
  File "/fsx/users/gcramer/work/pytorch/benchmarks/distributed/rpc/parameter_server/test.py", line 71, in run_trainer
    loss_fn(out, input.to(torch.float)).backward()
  File "/fsx/users/gcramer/work/pytorch/torch/_tensor.py", line 255, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/fsx/users/gcramer/work/pytorch/torch/autograd/__init__.py", line 149, in backward
    allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag
NotImplementedError: Cannot access storage of SparseTensorImpl
```

## Expected behavior

.backward() can access storage of SparseTensorImpl and gradients are updated
## Environment

```
Collecting environment information...
PyTorch version: 1.9.0a0+git7bc46bc
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.19.6

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB

Nvidia driver version: 450.80.02
cuDNN version: Probably one of the following:
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.9.0a0+git7bc46bc
[conda] magma-cuda110             2.5.2                         1    pytorch
[conda] mkl                       2021.2.0           h726a3e6_389    conda-forge
[conda] mkl-include               2021.2.0           h726a3e6_389    conda-forge
[conda] numpy                     1.19.2           py36h6163131_0  
[conda] numpy-base                1.19.2           py36h75fe3a5_0  
[conda] torch                     1.9.0a0+git7bc46bc           dev_0    <develop>
```

## Additional context

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @aazzolini @rohan-varma @jjlilley @osalpekar @jiayisuse @mrzzd @agolynski @SciPioneer @H-Huang @cbalioglu @gcramer23

**Comment by mrshenli** [comment]
cc @pritamdamania87 

@gcramer23 tried wrapping the hook implementation with `no_grad()` context, but hit the same error. How does autograd/distributed autograd work in this case?

**Comment by gcramer23** [comment]
cpp stack trace
```
On WorkerInfo(id=0, name=trainer):
NotImplementedError('Cannot access storage of SparseTensorImpl\nException raised from throw_storage_access_error at ../c10/core/TensorImpl.cpp:351 (most recent call first):\nframe #0: <unknown function> + 0x10a8d1 (0x7f1bdbfd68d1 in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)\nframe #1: std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > ()>::operator()() const + 0x4c (0x7f1c1e071250 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_cpu.so)\nframe #2: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7f1bdbfd5aaa in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)\nframe #3: <unknown function> + 0xe8c38 (0x7f1bdbfb4c38 in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)\nframe #4: c10::TensorImpl::throw_storage_access_error() const + 0xa0 (0x7f1bdbfb2206 in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)\nframe #5: <unknown function> + 0x211e0c1 (0x7f1c199ca0c1 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_cpu.so)\nframe #6: <unknown function> + 0x22dfad9 (0x7f1c19b8bad9 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_cpu.so)\nframe #7: c10::ivalue::Future::extractDataPtrs(c10::IValue const&) + 0x10f (0x7f1c19b850e5 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_cpu.so)\nframe #8: <unknown function> + 0x19fc856 (0x7f1c2d680856 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #9: <unknown function> + 0x1a1350f (0x7f1c2d69750f in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #10: <unknown function> + 0x1a7329d (0x7f1c2d6f729d in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #11: std::function<void (c10::ivalue::Future&)>::operator()(c10::ivalue::Future&) const + 0x49 (0x7f1c2d6a51a1 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #12: void c10::ivalue::Future::invokeCallback<std::function<void (c10::ivalue::Future&)> >(std::function<void (c10::ivalue::Future&)>) + 0x176 (0x7f1c2d693d0c in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #13: <unknown function> + 0x19fcce7 (0x7f1c2d680ce7 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #14: <unknown function> + 0x21fa62f (0x7f1c2de7e62f in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #15: <unknown function> + 0x21fe128 (0x7f1c2de82128 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #16: <unknown function> + 0x22023ea (0x7f1c2de863ea in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #17: std::function<void (c10::ivalue::Future&)>::operator()(c10::ivalue::Future&) const + 0x49 (0x7f1c2d6a51a1 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #18: void c10::ivalue::Future::invokeCallback<std::function<void (c10::ivalue::Future&)> >(std::function<void (c10::ivalue::Future&)>) + 0x176 (0x7f1c2d693d0c in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #19: <unknown function> + 0x19fcce7 (0x7f1c2d680ce7 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #20: <unknown function> + 0x2224a97 (0x7f1c2dea8a97 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #21: <unknown function> + 0x22279df (0x7f1c2deab9df in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #22: std::function<void ()>::operator()() const + 0x32 (0x7f1c2d7c2060 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)\nframe #23: c10::ThreadPool::main_loop(unsigned long) + 0x194 (0x7f1bdbfbcd3a in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)\nframe #24: <unknown function> + 0xf045f (0x7f1bdbfbc45f in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)\nframe #25: <unknown function> + 0xf1337 (0x7f1bdbfbd337 in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)\nframe #26: <unknown function> + 0xf1062 (0x7f1bdbfbd062 in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)\nframe #27: <unknown function> + 0xf1570 (0x7f1bdbfbd570 in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)\nframe #28: <unknown function> + 0xf152c (0x7f1bdbfbd52c in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)\nframe #29: <unknown function> + 0xf14fc (0x7f1bdbfbd4fc in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)\nframe #30: <unknown function> + 0xbd6df (0x7f1c3d6286df in /usr/lib/x86_64-linux-gnu/libstdc++.so.6)\nframe #31: <unknown function> + 0x76db (0x7f1c67ac36db in /lib/x86_64-linux-gnu/libpthread.so.0)\nframe #32: clone + 0x3f (0x7f1c677ec71f in /lib/x86_64-linux-gnu/libc.so.6)\n',)
Traceback (most recent call last):
  File "/fsx/users/gcramer/work/pytorch/torch/distributed/rpc/internal.py", line 210, in _run_function
    result = python_udf.func(*python_udf.args, **python_udf.kwargs)
  File "/fsx/users/gcramer/work/pytorch/benchmarks/distributed/rpc/parameter_server/test.py", line 72, in run_trainer
    loss_fn(ddp_model(input), input.to(torch.float)).backward()
  File "/fsx/users/gcramer/work/pytorch/torch/_tensor.py", line 255, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/fsx/users/gcramer/work/pytorch/torch/autograd/__init__.py", line 149, in backward
    allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag
NotImplementedError: Cannot access storage of SparseTensorImpl
Exception raised from throw_storage_access_error at ../c10/core/TensorImpl.cpp:351 (most recent call first):
frame #0: <unknown function> + 0x10a8d1 (0x7f1bdbfd68d1 in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)
frame #1: std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > ()>::operator()() const + 0x4c (0x7f1c1e071250 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_cpu.so)
frame #2: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x40 (0x7f1bdbfd5aaa in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)
frame #3: <unknown function> + 0xe8c38 (0x7f1bdbfb4c38 in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)
frame #4: c10::TensorImpl::throw_storage_access_error() const + 0xa0 (0x7f1bdbfb2206 in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)
frame #5: <unknown function> + 0x211e0c1 (0x7f1c199ca0c1 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_cpu.so)
frame #6: <unknown function> + 0x22dfad9 (0x7f1c19b8bad9 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_cpu.so)
frame #7: c10::ivalue::Future::extractDataPtrs(c10::IValue const&) + 0x10f (0x7f1c19b850e5 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_cpu.so)
frame #8: <unknown function> + 0x19fc856 (0x7f1c2d680856 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #9: <unknown function> + 0x1a1350f (0x7f1c2d69750f in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #10: <unknown function> + 0x1a7329d (0x7f1c2d6f729d in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #11: std::function<void (c10::ivalue::Future&)>::operator()(c10::ivalue::Future&) const + 0x49 (0x7f1c2d6a51a1 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #12: void c10::ivalue::Future::invokeCallback<std::function<void (c10::ivalue::Future&)> >(std::function<void (c10::ivalue::Future&)>) + 0x176 (0x7f1c2d693d0c in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #13: <unknown function> + 0x19fcce7 (0x7f1c2d680ce7 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #14: <unknown function> + 0x21fa62f (0x7f1c2de7e62f in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #15: <unknown function> + 0x21fe128 (0x7f1c2de82128 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #16: <unknown function> + 0x22023ea (0x7f1c2de863ea in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #17: std::function<void (c10::ivalue::Future&)>::operator()(c10::ivalue::Future&) const + 0x49 (0x7f1c2d6a51a1 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #18: void c10::ivalue::Future::invokeCallback<std::function<void (c10::ivalue::Future&)> >(std::function<void (c10::ivalue::Future&)>) + 0x176 (0x7f1c2d693d0c in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #19: <unknown function> + 0x19fcce7 (0x7f1c2d680ce7 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #20: <unknown function> + 0x2224a97 (0x7f1c2dea8a97 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #21: <unknown function> + 0x22279df (0x7f1c2deab9df in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #22: std::function<void ()>::operator()() const + 0x32 (0x7f1c2d7c2060 in /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_python.so)
frame #23: c10::ThreadPool::main_loop(unsigned long) + 0x194 (0x7f1bdbfbcd3a in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)
frame #24: <unknown function> + 0xf045f (0x7f1bdbfbc45f in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)
frame #25: <unknown function> + 0xf1337 (0x7f1bdbfbd337 in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)
frame #26: <unknown function> + 0xf1062 (0x7f1bdbfbd062 in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)
frame #27: <unknown function> + 0xf1570 (0x7f1bdbfbd570 in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)
frame #28: <unknown function> + 0xf152c (0x7f1bdbfbd52c in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)
frame #29: <unknown function> + 0xf14fc (0x7f1bdbfbd4fc in /fsx/users/gcramer/work/pytorch/torch/lib/libc10.so)
frame #30: <unknown function> + 0xbd6df (0x7f1c3d6286df in /usr/lib/x86_64-linux-gnu/libstdc++.so.6)
frame #31: <unknown function> + 0x76db (0x7f1c67ac36db in /lib/x86_64-linux-gnu/libpthread.so.0)
frame #32: clone + 0x3f (0x7f1c677ec71f in /lib/x86_64-linux-gnu/libc.so.6)
```

**Comment by gcramer23** [comment]
The data ptr extraction logic does not support sparse tensors.

**Comment by mrshenli** [comment]
closed by #59609
